### PR TITLE
5010 redesign product form

### PR DESCRIPTION
**Pull request by isaacfreeman**

### CODE REVIEW

File: backend/app/assets/javascripts/spree/backend/admin.js.erb [hunk 1]
```diff
@@ -123,7 +123,79 @@ handle_date_picker_fields = function(){
...
+function updateShowElementFromInput( editField ) {
+  var correspondingShowElement = '#show_' + editField.attr('id');
+  $(correspondingShowElement).html( editField.val() );
+}
+function updateShowElementFromSelect( editField ) {
+  var correspondingShowElement = '#show_' + editField.closest('.field-option_value').attr('id');
+  $(correspondingShowElement).html( editField.children('option:selected').text() );
+}
+
+function replaceId( element, newId ) {
```
Comment (djones) [inline backend/app/assets/javascripts/spree/backend/admin.js.erb]: Extra spaces around the parameters here. Other methods do not have this style.

File: backend/app/assets/javascripts/spree/backend/admin.js.erb [hunk 2]
```diff
@@ -123,7 +123,79 @@ handle_date_picker_fields = function(){
...
+  element.prop("id", element.prop("id").replace(/\d+/, newId))
```
Comment (djones) [inline backend/app/assets/javascripts/spree/backend/admin.js.erb]: double quotes used here when singles quotes is mostly used elsewhere

File: backend/app/assets/javascripts/spree/backend/admin.js.erb [hunk 3]
```diff
@@ -123,7 +123,79 @@ handle_date_picker_fields = function(){
...
+}
+function updateOptionTypes() {
+  var product_slug = $('form.edit_product').data('product-slug');
```
Comment (djones) [inline backend/app/assets/javascripts/spree/backend/admin.js.erb]: underscored variable names used here when camel case is used elsewhere.

File: backend/app/assets/javascripts/spree/backend/admin.js.erb [hunk 4]
```diff
@@ -123,7 +123,79 @@ handle_date_picker_fields = function(){
...
+  var option_type_ids = $('#product_option_type_ids').val();
+  $.ajax({  dataType: 'json',
+            url: "/admin/products/" + product_slug,
+            data: { product: { option_type_ids: option_type_ids }},
+            type: 'patch',
+            success: function( data ) {
```
Comment (djones) [inline backend/app/assets/javascripts/spree/backend/admin.js.erb]: Spaces around parameters. I won't comment on this any further on this as there are a number of examples. Pick a style and stick with it. Same with the camel case vs underscored.

File: backend/app/assets/javascripts/spree/backend/admin.js.erb [hunk 5]
```diff
@@ -154,11 +228,47 @@ $(document).ready(function(){
     // When cloning a new row, set the href of all icons to be an empty "#"
     // This is so that clicking on them does not perform the actions for the
     // duplicated row
-    new_table_row.find("a").each(function () {
+    newElement.find("a").each(function () {
       var el = $(this);
       el.prop('href', '#');
```
Comment (djones) [inline backend/app/assets/javascripts/spree/backend/admin.js.erb]: this could just be:

```
$(this).prop('href', '#');
```

File: backend/app/assets/javascripts/spree/backend/admin.js.erb [hunk 6]
```diff
@@ -154,11 +228,47 @@ $(document).ready(function(){
...
+    });
+    newVariant.find(".field-option_value").each(function () {
+      replaceId( $(this), newId);
+    });
+
+
+    // When cloning a new row, set the href of all icons to be an empty "#"
+    // This is so that clicking on them does not perform the actions for the
+    // duplicated row
+    newVariant.find("a").each(function () {
```
Comment (djones) [inline backend/app/assets/javascripts/spree/backend/admin.js.erb]: Looks like a bit of duplication here with a several lines above. Anyway to put this stuff related to cloning in a reusable method?

File: backend/app/assets/javascripts/spree/backend/admin.js.erb (hunk 5 above)
Comment (isaacfreeman) [inline backend/app/assets/javascripts/spree/backend/admin.js.erb]: Yeah, I changed that when I made a special version of this specific to variants, but I guess I've mucked around with this generic version enough to say I own it too.

File: backend/app/assets/javascripts/spree/backend/admin.js.erb [hunk 7]
```diff
@@ -271,6 +381,33 @@ $(document).ready(function(){
...
+          ui.placeholder.height(ui.item.height())
+          // Fix placeholder content to make it correct width
+          ui.placeholder.html("<td colspan='"+(td_count-1)+"'></td><td class='actions'></td>")
+        },
+        stop: function (event, ui) {
+          // Fix odd/even classes after reorder
+          $("table.sortable tr:even").removeClass("odd even").addClass("even");
+          $("table.sortable tr:odd").removeClass("odd even").addClass("odd");
+        }
+
```
Comment (djones) [inline backend/app/assets/javascripts/spree/backend/admin.js.erb:586]: extra line here

File: backend/app/assets/stylesheets/spree/backend/sections/_products.scss [hunk 8]
```diff
@@ -1,3 +1,10 @@
+form.edit_product {
```
Comment (djones) [inline backend/app/assets/stylesheets/spree/backend/sections/_products.scss:1]: You're not making use of the nesting here so you could just make this:

```
form.edit_product .inline-menu {
  float: right;
  margin-bottom: 10px;
}
```

File: backend/app/assets/javascripts/spree/backend/admin.js.erb (hunk 6 above)
Comment (isaacfreeman) [inline backend/app/assets/javascripts/spree/backend/admin.js.erb]: Yeah, I'll have another look at that. It's also violating my principle of keeping jQuery handler functions as short as possible. It got this way because I went back and forth on whether to write a new function for variants or try to modify the existing one. Will refactor both.

File: backend/app/assets/stylesheets/spree/backend/sections/_products.scss [hunk 9]
```diff
@@ -35,6 +60,97 @@
...
+    }
+    tr.edit {
+      td {
+        vertical-align: top;
+        border-right: none;
+        border-top: 1px solid $color-border;
+        &:nth-child(6) {
+          border-right: 1px solid $color-border;
+        }
+
```
Comment (djones) [inline backend/app/assets/stylesheets/spree/backend/sections/_products.scss]: extra line here

File: backend/app/assets/stylesheets/spree/backend/sections/_products.scss [hunk 10]
```diff
@@ -35,6 +60,97 @@
...
+  }
+  tbody.ui-sortable-placeholder {
+    display: table-row;
+    td {
+      border: 1px solid $color-2 !important;
+      visibility: visible !important;
+
+      &.actions {
+        background-color: transparent;
+        border-right: none !important;
```
Comment (djones) [inline backend/app/assets/stylesheets/spree/backend/sections/_products.scss]: You could make this:

```
border: none !important;
border-left: 1px solid $color-2 !important;
```

Not sure if that's weird though.

File: backend/app/views/spree/admin/products/_form.html.erb [hunk 11]
```diff
@@ -177,3 +119,22 @@
   $('.select2-container').css({width: '20em'})
   </script>
 <% end %>
+
+
+<script>
```
Comment (djones) [inline backend/app/views/spree/admin/products/_form.html.erb]: this'd be better in a JS file somewhere...

File: backend/app/assets/stylesheets/spree/backend/sections/_products.scss (hunk 10 above)
Comment (isaacfreeman) [inline backend/app/assets/stylesheets/spree/backend/sections/_products.scss]: Not weird at all.

You can tell this is copy-and-pasted code because it's got !important in it. These styles should really be in the shared/_tables.scss, with more generic class names.

File: backend/app/views/spree/admin/products/_form.html.erb [hunk 12]
```diff
@@ -177,3 +119,22 @@
...
+  $( "a#manage_properties" ).click(function( event ) {
+    event.preventDefault();
+    $( '#admin_product_form_product_properties' ).slideToggle();
+  });
+  $( "a#add_variants" ).click(function( event ) {
+    event.preventDefault();
+    $( '#admin_product_form_multiple_variants' ).slideToggle();
```
Comment (djones) [inline backend/app/views/spree/admin/products/_form.html.erb]: I think you may be able to shorten this to just:

```
$( '#admin_product_form_multiple_variants, #admin_product_form_master_variant' ).slideToggle();
```

http://api.jquery.com/multiple-selector/

File: backend/app/views/spree/admin/products/_form.html.erb (hunk 11 above)
Comment (isaacfreeman) [inline backend/app/views/spree/admin/products/_form.html.erb]: Another case of "should I fix this?" And again, I've changed this file so much now that there's no point in still being reluctant to move this.

File: backend/app/views/spree/admin/products/_product_properties.html.erb [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    <% end %>
+  </tbody>
+</table>
+
+<%= javascript_tag do -%>
+  var properties = <%= raw(@properties.to_json) %>;
+  $('#product_properties').on('keydown', 'input.autocomplete', function() {
+    already_auto_completed = $(this).is('ac_input');
+    if (!already_auto_completed) {
+      $(this).autocomplete({source: properties});
```
Comment (djones) [inline backend/app/views/spree/admin/products/_product_properties.html.erb]: might be able to chain this.

```
$(this).autocomplete({source: properties}).focus();
```

File: backend/app/views/spree/admin/products/_product_properties.html.erb [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    </tr>
+  </thead>
+  <tbody id="product_properties" data-hook>
+    <%= f.fields_for :product_properties do |pp_form| %>
+      <%= render 'spree/admin/product_properties/product_property_fields', :f => pp_form %>
+    <% end %>
+  </tbody>
+</table>
+
+<%= javascript_tag do -%>
```
Comment (djones) [inline backend/app/views/spree/admin/products/_product_properties.html.erb]: Would be good to put this in a JS file somewhere.

At very least this code doesn't wait until the dom is ready but it probably should do?

File: backend/app/views/spree/admin/products/_seo_fields.html.erb [hunk 15]
```diff
@@ -0,0 +1,13 @@
+<div data-hook="admin_product_form_meta_keywords">
```
Comment (djones) [inline backend/app/views/spree/admin/products/_seo_fields.html.erb:1]: Duplication here. Make it so you can just pass in the field names `meta_keywords` and `meta_description` to have it render the same HTML.

File: backend/app/views/spree/admin/prototypes/select.js.erb [hunk 16]
```diff
@@ -1,4 +1,5 @@
+<%# binding.remote_pry %>
```
Comment (djones) [inline backend/app/views/spree/admin/prototypes/select.js.erb]: remove this line

File: backend/app/views/spree/admin/variants/_master_variant_fields.html.erb [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+  <div data-hook="admin_product_form_sku">
+    <%= f.field_container :sku do %>
+      <%= f.label :sku, Spree.t(:sku) %>
+      <%= f.text_field :sku, :size => 16 %>
+    <% end %>
+  </div>
+</div>
+<div class="six columns">
+  <ul id="shipping_specs">
+    <li id="shipping_specs_weight_field" data-hook="admin_product_form_weight" class="field alpha omega six columns">
```
Comment (djones) [inline backend/app/views/spree/admin/variants/_master_variant_fields.html.erb]: Some duplication over weight, height, width and depth. May not be worth putting into a partial though.

File: backend/app/views/spree/admin/products/_seo_fields.html.erb (hunk 15 above)
Comment (isaacfreeman) [inline backend/app/views/spree/admin/products/_seo_fields.html.erb:1]: Disagree on this one. Having two separate groups with data-hooks makes it much easier to use Deface to modify the form. Having just had to deal with using Deface on views that didn't account for this in spree_static_content, this seems more important to me than being DRY.

File: backend/app/views/spree/admin/variants/_master_variant_fields.html.erb (hunk 17 above)
Comment (isaacfreeman) [inline backend/app/views/spree/admin/variants/_master_variant_fields.html.erb]: Another one that's there for Deface, so it's easy to get at the particular fields.

I do find these dimension fields weird, as they're almost always left empty. I guess they're needed for some shipping methods, though, so they need to be present. It's a pain on the front end, because they take up most of the variant form with little value to most users. Still, that's another UX question for another day. I'm trying to keep focused on overall variant management for this PR.

File: backend/app/views/spree/admin/variants/_variant.html.erb [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+      <td class="no-border">
+        <span class="handle"></span>
+        <%= vf.hidden_field :position, :value => variant.position, :class => 'position_field' %>
+      </td>
+      <td id="show_product_variants_attributes_<%= variant_counter %>_sku" class="field-display"><%= variant.sku %></td>
+      <% @product.option_types.each do |option_type| %>
+        <td id="show_variant_<%= variant_counter %>_option_<%= option_type.name %>" class="option_value field-display"><%= variant.option_value(option_type.name) %></td>
+      <% end %>
+      <td>
+        <span id="show_product_variants_attributes_<%= variant_counter %>_height" class="field-display">
```
Comment (djones) [inline backend/app/views/spree/admin/variants/_variant.html.erb]:
``` erb
<% ['height', 'width', 'depth'].each do |measurement| %>
  <span id="show_product_variants_attributes_<%= variant_counter %>_<%= measurement %>" class="field-display">
    <%= number_with_precision(variant.send(measurement), :precision => 2) %>
  </span><%= measurement.chars.first %>
  &times;
<% end %>
```

File: core/app/models/spree/product.rb [hunk 19]
```diff
@@ -93,6 +93,8 @@ class Product < Spree::Base
 
     accepts_nested_attributes_for :product_properties, allow_destroy: true, reject_if: lambda { |pp| pp[:property_name].blank? }
 
+    accepts_nested_attributes_for :variants, allow_destroy: true, :reject_if => lambda { |v| v[:sku].blank? }
```
Comment (djones) [inline core/app/models/spree/product.rb:96]: Ruby 1.9 syntax on `allow_destroy: true`. I think you've used mostly 1.8 elsewhere.

File: backend/app/views/spree/admin/variants/_variant.html.erb (hunk 18 above)
Comment (isaacfreeman) [inline backend/app/views/spree/admin/variants/_variant.html.erb]: Another case for Deface. 

Dimensions don't have any units, and I suspect that adding them in with Deface would be a common thing to do (if it isn't it should be). So really these need to be kept separate and have data-hooks added to them.

I do need to review this though, because it doesn't work on the front end – if there are no dimensions you get "h x w x d" with no numbers.
